pkg/endpoint/api/endpoint_deletion_queue.go

```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package api
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"os"
+	"path/filepath"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/job"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/api/v1/server"
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/endpointstate"
+	"github.com/cilium/cilium/pkg/lock/lockfile"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/promise"
+)
+
+type DeletionQueue struct {
+	logger                 *slog.Logger
+	lf                     *lockfile.Lockfile
+	endpointRestorePromise promise.Promise[endpointstate.Restorer]
+	endpointAPIManager     EndpointAPIManager
+	processed              chan struct{}
+}
+
+func (dq *DeletionQueue) Process(ctx context.Context, health cell.Health) error {
+	if _, err := dq.endpointRestorePromise.Await(ctx); err != nil {
```

Comment:
I wonder whether locking & processing the deletion queue needs to wait for full endpoint restoration (calling `Restorer.WaitForEndpointRestore` in addition to waiting for the promise)? Otherwise the deletions might (silently) fail because they aren't yet added/exposed via the `EndpointManager`?

Note: Until now we waited for the `Daemon` instead of the `Restoration` promise - but we didn't wait for full endpoint restoration either.

@squeed :eyes:

Reply:
As I mentioned, the deletion queue didn't wait for the full restoration to succeed before this PR either - therefore i think it's ok to fix this in a follow-up PR if this is an actual issue.